### PR TITLE
Performance: Add a tool and a CI workflow to compare performance between PR and trunk and track metrics on trunk

### PR DESCRIPTION
**Pull request by youknowriad**
### Changes proposed in this Pull Request:

**Big Picture** My goal is to bring metrics tracking and performance dashboard to the tools WooCommerce use daily to avoid performance regressions and work on performance improvements similarly to the [Gutenberg project](https://www.codevitals.run/project/gutenberg) for instance.

Implementing this require three complementary steps:

 1- First step is to have some test runner (e2e test here) that we can launch to compute the metrics.
 2- Second step is to be able to use the same test runner in two separate branches as part of the same CI job. The reason for this is that CI runners capacities are unstable and we can't rely on the absolute numbers. We have to compute relative numbers and compare each commit with a given "base commit".
 3- Third step is to send these values (base branch and commit) to the dashboard (codevitals) to log it and produce the graphs.

The first step is being implemented in https://github.com/woocommerce/woocommerce/pull/42005
The current PR implements the second step. (It also includes the commits from the first step for now until it's merged)

So it adds a "compare-perf" tool that was copied and adapted from the Gutenberg repository. It's almost a generic tool that can be used in any plugin repo. the only specific things at the moment is the `config.js` file. So maybe down the road, we'll have that tool as a reusable package in Gutenberg but for now this should work well.

The PR also adds a "metrics" workflow to CI. The workflow runs for each PR and compare its metrics with trunk. And it also runs on trunk for each commit and compares the commit with a "fixed" point of comparison (this is going to be used later to track the number on a graph like [here](http://codevitals.run)) 

### How to test the changes in this Pull Request:

You can run the perf tool locally like so (give random commits or tags to compare, the first commit must include the perf tests though as it's going to be used to run the tests)

```
cd tools/compare-perf
pnpm test perf 2f6ca66e00b3666b2567877ae67cbfb5b6ce171a  61438bc23746eb78700d87d2cabcdae8cf10de30
```

you'll get something close to the following results:

Screenshot 2023-12-12 at 10 40 51 AM

Also, we can check the results of the "metrics" workflow on the PR as well.

### CODE REVIEW

File: plugins/woocommerce/tests/metrics/specs/editor.spec.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+				props.preferences
+			) ) {
+				await window.wp.data
+					.dispatch( 'core/preferences' )
+					.set( props.context, key, value );
+			}
+		},
+		{ context, preferences }
+	);
+}
```
Comment (WunderBart) [inline plugins/woocommerce/tests/metrics/specs/editor.spec.js]: I assume these haven't been released with the core package yet. In such case, we can still extend the original fixture by doing the following:

```js
import { Editor } from '@wordpress/e2e-test-utils-playwright';

class EditorExtended extends Editor {
	async editPost( postId ) {
		// ...
	}

	async setPreferences( context, preferences ) {
		// ...
	}
}

test.use( {
	editor: async ( { page }, use ) => {
		await use( new EditorExtended( { page } ) );
	},
} );
```

This way, we don't need to update the tests.

Comment (youknowriad) [reply plugins/woocommerce/tests/metrics/specs/editor.spec.js]: They have been released but there's some rules I guess in the repo about "fixing dependencies". I didn't want to update the dependency as it's used elsewhere and I don't know the consequences. But yeah, good call about the extension, I didn't know about it :)

**Comment by youknowriad** [comment]
Screenshot 2023-12-14 at 11 45 53 AM

Good news, the metrics job is running perfectly and the initial results seem stable enough.
It took 29mn to run the job. It takes 36mn in Gutenberg with a few more metrics, I think that's acceptable for a start. We can still optimize though. 

I'm going to attempt an optimization (not sure it will work) but if it doesn't we can revert my next commit :P

**Review by senadir — APPROVED** [review]
All tests are passing and I'm seeing results! Would love to get the time to be less but that's not a huge blocker.

**Comment by senadir** [comment]
Took 24mn in this optimization so that's good for me!

**Pull request merged**